text/0000-placement-by-return.md

```diff
@@ -0,0 +1,1110 @@
+
+In the case of array and struct literals, the copy elision must be recursive. For example:
+
+```rust
+fn no_copy() -> Struct2 {
+    Struct2 { member: Struct { member: getData() } }
+}
+```
+
+In the above example, the data returned by `getData()` may be sized or unsized. Either way, it should not be copied at all (aside from its initial creation), whether when constructing the Struct, the Struct2, or returning the data.
+
+Guaranteed Copy Elision only kicks in when a GCE-applicable expression is directly returned from a function, either by being the function's last expression or by being the expression part of a `return` statement.
+
+## The unsafe method of writing return values
+
+The `std::mem` module exposes two symmetrical intrinsics that can allow you to achieve Copy Elision in any circumstance.
+
+### `write_return_with`
+
+    pub unsafe fn write_return_with<T, F: for<'a> FnOnce(*mut u8)>(f: F) -> T {
+        write_unsized_return_with(Layout::new::<T>(), |p| {f(p); &mut *(p as *mut T)})
+    }
+    extern "intrinsic" pub unsafe fn write_unsized_return_with<T: ?Sized, F: for<'a> FnOnce(*mut u8) -> &mut T>(layout: Layout, f: F) -> T;
+
+[write_return_with]: #write_return_with
+
+Directly access the caller-provided return slot.
+
+The return slot is an uninitialized memory space provided by a function's caller to write the return value, and implicitly passed as a pointer. A function which directly returns the result of evaluating another function, like `fn foo() -> ReturnedValue { bar() }`, may simply pass the slot along, thus avoiding expensive copies. In cases where it is impossible to implement a function by directly returning the value of a function call or a primitive expression, directly accessing the return slot and writing it directly may be necessary.
+
+Return slots are not always used; small values like pointers, numbers, and booleans may be returned by filling a register instead. In this case, `write_return_with` will implicitly create a "return slot", pass a pointer to it to `f`, and then copy the return value from the "return slot" to the register.
+
+The pointer returned by `write_unsized_return_with`'s callback should be the same as the one provided to it, along any necessary unsize information (such as the slice's length or the trait object's vtable pointer). Additionally, when using `write_unsized_return_with` for sized types, the provided layout must be exactly the same as the one produced by `Layout::new<T>()`. These two values must be correct even when they're redundant. A function that returns a slice may return a slice that is smaller than the requested allocation Layout, in case it is unable to predict the amount of data that will be available to it, but when producing a trait object, it must know exactly the right size for its allocation.
```

Comment:
> A function that returns a slice may return a slice that is smaller than the requested allocation Layout

How is the caller supposed to deal with this? This seems not possible in the case of `Box::<[T]>::new_with`, it would need to re-allocate (and therefore possibly copy) in order to support the smaller size. I believe your implementation of `Drop for Box` below is going to call `System.dealloc` with an incorrect layout.

Reply:
It has to copy.

This exists for the benefit of high-performance input/output; what happens when we allocate a buffer for zero-copy I/O, and then it fails when we actually try to run the operation? If you're able to get an accurate minimum bound for available data, then you will either return the full-sized slice (which means no copying) or an empty sentinel slice (which also means no copying).

But as a general interface, this is a performance cliff. The example implementation of `Box` also needs updated to account for it.

Reply:
I think it'd be better to forbid this and instead fail whole thing but it requires fallibility to be addressed.